fml/synchronization/sync_switch.h

```diff
@@ -0,0 +1,54 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#ifndef FLUTTER_FML_SYNCHRONIZATION_SYNC_SWITCH_H_
+#define FLUTTER_FML_SYNCHRONIZATION_SYNC_SWITCH_H_
+
+#include <forward_list>
+#include <functional>
+#include <mutex>
+#include "flutter/fml/macros.h"
+
+namespace fml {
+
+/// A threadsafe structure that allows you to switch between 2 different
+/// execution paths.
+class SyncSwitch {
+ public:
+  /// Represents the 2 code paths available when calling |SyncSwitch::Execute|.
+  struct Handlers {
+    /// Creates a |Handlers| were all exection paths are noops.
+    Handlers();
+    Handlers& SetTrue(const std::function<void()>& handler);
```

Comment:
Why doesn't it make sense?  This is avoiding an extra copy.

Reply:
Since you are passing it here by const reference but storing by value, the store into the ivar would guarantee a copy. So you didn't quite avoid it. If you really wanted to avoid copies, you would pass it by value here and move it into the ivar. The caller can then also move it into the invocation as well to avoid potential copies there. But the functor is just a callable with some state for the captures so it is pretty small and pass by value is good enough. In either case, the const reference adds nothing. OTOH, if you are are passing by const reference and storing by const reference, then I suppose it makes sense. But that makes it incumbent on the caller to keep the thing alive which is not always guaranteed to happen. I don't know exactly how many words it is for the a functor, but pass by value is fast and safe.

Reply:
Right, 2 copies versus 1: [variable -> argument -> ivar] vs [variable -> ivar].  I don't see a reason to introduce a superfluous copy when it's easy to just pass by reference.  I've been docked for this during official google c++ code reviews.

Reply:
The "extra" copy is 6 words on a 64-bit system. If you absolutely must know, the code generated by the compiler is exactly the same https://godbolt.org/z/fBkPDJ.

> I've been docked for this during official google c++ code reviews.

While this is unfortunate, we only have the style guide to go by (at go/fluttercpp). I could not find guidance for the preference of how to pass functors as parameters. If anything the guidance in parting thoughts about being consistent with the rest of the codebase might be more applicable.

Reply:
I'll entertain the thought a bit further strictly for fun.  I don't see a compelling reason to change it =)

If you turn off "-Os" and add a caller to `foo` you'll see there is a copy constructor used at the call site and an operator= inside the setter.  The size of the copy isn't bound to 6 words because it includes variables captured by the closure and it has a CPU cost in operations and loss of locality.  Avoiding any divergence between debug builds and release builds should be a goal when it isn't onerous.

It's a dangerous game to look into the output of compilers and let that inform your code since it isn't dictated by the language.   Are we using clang on windows or msvc++?  You'd have to assert that optimization on that compiler as well (although I don't think we should since we have a better solution).

Reply:
I added the compiler output only because the assertion was that the fewer copies were worth the const reference. I don't expect anyone to refer to compiler output.

The compelling reason ought to be that this deviates from convention used throughout the engine. There is no guidance related to this in the style guide and this convention was used long before you or I started working on the engine. If the guidance is going to "follow the style guide but do this one thing this way because the author got docked in a code review for it once", then that is not sustainable.

> I'll entertain the thought a bit further strictly for fun.

I assure you this is very much not fun for me. But I must insist.

Reply:
I hope you don't take it personal, we are just working out a minor C++ detail.

I'd say style arguments are reserved for matters of style, i.e. matters that don't have an affect on the outputted binary.  How about this, we'll just delegate to @cbracken.  Does consistency in usage of references in parameters trump performance?  Sound good?  We obviously would chose differently if given the option but I think we both respect @cbracken's take.

Reply:
What if we keep this consistent with the rest of the code in the engine, and then if we're convinced there's a performance gain to be had from changing it we can change all the callsites at once separately?

Reply:
We seem to be at an impasse so a third party to mediate this sounds good. Either way (and as I stated in my original comment) this is a nit. I am not going to stop anyone from committing this based on just this deviance from convention. If you feel really strongly about this, let me know if you'd rather I just review the other parts of this patch.

Reply:
I switched to rvalues, give it a look!